### PR TITLE
SinglePlayerPuzzles Part1: Data migration

### PR DESCRIPTION
**Pull request by digit01Wave**
#720 (part 1)
We want to introduce a concept of a SinglePlayerPuzzle, which can be solved as an individual player rather than as a team.
The player does not have to be on a team.

This PR is the first part that adds the necessary new tables. These parts generally split up this massive code change (https://github.com/PuzzleServer/mainpuzzleserver/pull/792)

This PR adds the following 4 tables to the database:
- `SinglePlayerPuzzleHintStatePerPlayer` - every row is the hint state of a specific player for a specific single player puzzle
- `SinglePlayerPuzzleStatePerPlayer` - every row is the puzzle state (solve time, lockout, etc.) of a specific player for a specific single player puzzle
- `SinglePlayerPuzzleSubmission` - every row is is a submission to a single player puzzle
- `SinglePlayerPuzzleUnlockState` - every row is the unlock state of a given SinglePlayerPuzzle. <br />Note that this is different from the regular PuzzleStatePerTeam that has an unlock property that the author/admin can set to manually unlock the puzzle for a specific team. When it comes to single player puzzles, the author/admin will only be able to unlock ALL or lock ALL. This is because we don't want to keep adding all users to the SinglePlayerPuzzleStatePerPlayer every time someone new logs in.

### CODE REVIEW

File: Data/Migrations/PuzzleServerContextModelSnapshot.cs [hunk 1]
```diff
@@ -15,9 +17,13 @@ protected override void BuildModel(ModelBuilder modelBuilder)
         {
 #pragma warning disable 612, 618
             modelBuilder
-                .UseIdentityColumns()
-                .HasAnnotation("Relational:MaxIdentifierLength", 128)
-                .HasAnnotation("ProductVersion", "5.0.2");
+                .HasAnnotation("ProductVersion", "7.0.2")
+                .HasAnnotation("Proxies:ChangeTracking", false)
```
Comment (morganbr) [inline Data/Migrations/PuzzleServerContextModelSnapshot.cs:21]: Please coordinate with @asyasky on this and #790

File: ServerCore/Pages/Submissions/Index.cshtml [hunk 2]
```diff
@@ -98,7 +98,7 @@
                 <span>You've submitted too many wrong responses. Please <a href="@mailtoUrl">use this link to email us your answer for @Model.Puzzle.PlaintextName or ask for help</a>.</span>
             </div>
         }
-        else if (Model.PuzzleState.IsTeamLockedOut)
+else if (Model.PuzzleState.IsLockedOut)
```
Comment (morganbr) [inline ServerCore/Pages/Submissions/Index.cshtml:101]: Nit: It looks like something happened to indentation here

**Review by morganbr — APPROVED** [review]
Looks good once the EF migration questions are solved

File: ServerCore/Pages/Submissions/Index.cshtml (hunk 2 above)
Comment (digit01Wave) [reply ServerCore/Pages/Submissions/Index.cshtml]: fixed in newest iteration

File: Data/DataModel/SinglePlayerPuzzleUnlockState.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace ServerCore.DataModel
+{
+    /// <summary>
+    /// The unlock state of a single player puzzle.
+    /// Unlike puzzles for teams, which allows admins/authors to unlock the puzzle for a specific team,
+    /// they will only be able to either lock or unlock for ALL players.
```
Comment (tabascq) [inline Data/DataModel/SinglePlayerPuzzleUnlockState.cs:9]: How will we test these puzzles then?

Comment (tabascq) [reply Data/DataModel/SinglePlayerPuzzleUnlockState.cs:9]: Will we treat them as standard team puzzles and switch them over late in the process?

Comment (digit01Wave) [reply Data/DataModel/SinglePlayerPuzzleUnlockState.cs:9]: So. I was thinking that we would lock the puzzle behind a pre-requisite and we can mark as solved the prerequisite to do this. A little hacky but better than constantly checking for every player that logs in.

Comment (digit01Wave) [reply Data/DataModel/SinglePlayerPuzzleUnlockState.cs:9]: Another way of doing it is to add a button that manually adds a player to the SinglePlayerPuzzleStatePerPlayer and allowing unlocks the typical way. Which I actually like more actually

Comment (tabascq) [reply Data/DataModel/SinglePlayerPuzzleUnlockState.cs:9]: If the way I suggested is less work then I think it's fine as all puzzles will start their lives that way.

Comment (digit01Wave) [reply Data/DataModel/SinglePlayerPuzzleUnlockState.cs:9]: sounds good to me!

**Pull request merged**